src/zdns/dnssec_types.go

```diff
@@ -0,0 +1,160 @@
+/*
+ * ZDNS Copyright 2024 Regents of the University of Michigan
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+package zdns
+
+import (
+	"context"
+
+	"github.com/miekg/dns"
+)
+
+// DNSSECStatus represents the overall validation status according to RFC 4035
+type DNSSECStatus string
+
+const (
+	DNSSECSecure        DNSSECStatus = "Secure"
+	DNSSECInsecure      DNSSECStatus = "Insecure"
+	DNSSECBogus         DNSSECStatus = "Bogus"
+	DNSSECIndeterminate DNSSECStatus = "Indeterminate"
+)
+
+type RRsetKey Question
+
+func (r *RRsetKey) String() string {
+	return "name: " + r.Name + ", type: " + dns.TypeToString[r.Type] + ", class: " + dns.ClassToString[r.Class]
+}
+
+// DNSSECPerSetResult represents the validation result for an RRSet
+type DNSSECPerSetResult struct {
+	RRset     RRsetKey     `json:"rrset"`
+	Status    DNSSECStatus `json:"status"`
+	Signature *RRSIGAnswer `json:"sig"`
+	Error     string       `json:"error"`
+}
+
+// DNSSECResult captures all information generated during a DNSSEC validation
+type DNSSECResult struct {
+	Status        DNSSECStatus         `json:"status" groups:"dnssec,dnssec,normal,long,trace"`
+	DS            []*DSAnswer          `json:"ds" groups:"dnssec,long,trace"`
+	DNSKEY        []*DNSKEYAnswer      `json:"dnskey" groups:"dnssec,long,trace"`
+	Answer        []DNSSECPerSetResult `json:"answer" groups:"dnssec,long,trace"`
+	Additionals   []DNSSECPerSetResult `json:"additionals" groups:"dnssec,long,trace"`
```

Comment:
I think this should be `Additional`  not `Additionals` per RFC and in matching the other things.

Reply:
We currently have `json:"additionals` in `SingleQueryResult` 
https://github.com/zmap/zdns/blob/d403578a6ca4aaed047ab6bc452e67b7df4a9419/src/zdns/qa.go#L77-L79

and in `CachedResult`:
https://github.com/zmap/zdns/blob/d403578a6ca4aaed047ab6bc452e67b7df4a9419/src/zdns/cache.go#L35-L38

I can address the pluralization issue for DNSSEC and cache (ideally in a separate PR). However, for `SingleQueryResult`, since the plural form is embedded in the JSON schema, changing it could introduce backward compatibility issues. 

If we leave it unchanged, we’ll have a mismatch where the JSON schema for DNSSEC uses the singular form, but the response retains the plural form. This inconsistency might be confusing.

To resolve this, I suggest we standardize by either:
1. Removing all plural forms and break compatibility, or 
2. Consistently using plural forms, at least for the JSON output.

What do you think?

Reply:
We're about to do a major 2.0 release. I think we should have it consistent everywhere, even if it means that schema changes.

Reply:
Fixed in dnssec code. Created #484 for other places.